### ISSUE TITLE
Investigate "eaten exceptions" in background task

### ISSUE THREAD

**Issue by amiller** [initc3/HoneyBadgerMPC#125]
We have already been using a workaround for exceptions in background tasks, which dont' seem to get printed by default. See #39.

However, when I try to reproduce this in some simple example, it doesn't happen!
The workarounds include: 
- explicitly attaching a callback with `add_done_callback` to background tasks, 
- installing `loop.set_exception_handler` to a handler that prints the exception

Furthermore, we noticed that exceptions in the `client_connection_cb` callback function passed to `asyncio.create_server` would be suppressed. #124  

However, this also is not easily reproduced in a simple example. So, something seems to be going wrong in our understanding.  

To wit:
```python3
import asyncio

HOST = 'localhost'
PORT = 8080

async def handle_client(reader, writer):
    print('handle_client')

    # Expected behavior:                                                                                                                                                                      
    #    without this exception, everything completes                                                                                                                                         
    # with this exception, it hangs                                                                                                                                                           
    #   In this example, the exception is printed, as I'd hope                                                                                                                                
    #   But, in our application, the exception is supppressed                                                                                                                                 
    #      *confused*                                                                                                                                                                         
    raise Exception('Oh no')

    writer.write(b'hi')
    print('handle_client done')

def server():
    print('server')
    server = asyncio.start_server(handle_client, HOST, PORT)
    return asyncio.ensure_future(server)

async def client():
    print('client')
    await asyncio.sleep(0.1)
    reader, writer = await asyncio.open_connection(HOST, PORT)
    print('client done:', await reader.read(2))

async def go():
    s = server()
    c = asyncio.ensure_future(client())
    await asyncio.sleep(1)
    return # Don't even clean up the server, will it print errors?                                                                                                                            

if __name__ == '__main__':
    asyncio.set_event_loop(asyncio.new_event_loop())
    loop = asyncio.get_event_loop()
    loop.run_until_complete(go())
```

**Comment by amiller** [comment]
I have a suspicion of what could explain this behavior. Going to describe it in this comment, so we can later check to see if it works.
 In our IPC code, each `handle_client` creates a while loop, which stays open until it is explicitly cancelled. In our `close()` function, we go through each such task and cancel it.
https://github.com/initc3/HoneyBadgerMPC/blob/5eab8b5126f2be63f244befd8e18f0a852a125fc/honeybadgermpc/ipc.py#L171
 However, notice that we do not subsequently `await` on the task after calling `task.cancel()`.
 
What we noticed elsewhere #149 is that the `handle_client` task creates other background tasks with `ensure_future`, and when `handle_client` is cancelled, the cancellation propagates to other tasks, which then have to be awaited on. We fixed this by cancelling the client and awaiting on it, except for the cancellation. However, this still doesn't quite explain the eaten exception messages, though.
